CIPs/CIP-124/cip-124.md

```diff
@@ -0,0 +1,321 @@
+type InterestRange struct {
+	start EventID // A string less the any event of intrest
+	dividers [
+    {
+      associative_hash string // the associative hash of all EventIDs in sub-range
+      link CID // [optional] the CID of the subnode
+
+      // an EventID larger then any EventID in the preceding sub-range
+      // and less then any in the subsuquent sub-range
+      stop string 
+
+    }
+  ]
+}
+
+type Interests: InterestRange[]
+```
+
+#### **Interactive**
+One or more interest range,
+each containing one or more sub-tree associative hashs.
+```
+(EventID (sub-tree-ahash EventID)+ )+
+```
+We folow a syncronisation protocal inspired by
+"Range-Based Set Reconciliation and Authenticated Set Representations"
+[[arXiv:2212.13567](https://doi.org/10.48550/arXiv.2212.13567)]
+
+---
+![ring](assets/ring.png)
+You initate a syncronisation by sending a hash covering the entire range.
+```js
+their_events: {"bee", "cat", "doe", "eel", "fox", "hog"}
+your_events:  {"ape", "eel", "fox", "gnu"}
+
+you: -["ape", h("eel", "fox", "gnu"), "ape"]->
+```
+They don't have the same hash so they split the range near the middel at `"eel" `
+```js
+their_events: {"ape", "bee", "cat", "doe", "eel", "fox", "hog"}
+your_events:  {"ape", "eel", "fox", "gnu"}
+
+them: <-["ape", h("bee", "cat", "doe"), "eel", h("fox", "hog"), "ape"]-
+```
+You have nothing from `"ape"` to `"eel"` so send `0`.
+Your hash from `"eel"` to `"ape"` dose not match so split range on `"gnu"`
+```js
+their_events: {"ape", "bee", "cat", "doe", "eel", "fox", "hog"}
+your_events:  {"ape", "eel", "fox", "gnu"}
+
+you: -["ape", 0, "eel", h("fox"), "gnu", 0, "ape"]->
+```
+They see that you don't have any keys form `"ape"` to `"eel"` so they send the entire set.
+
+They have a difrent hash from `"gnu"` to `"ape"` and split the range on `"hog"`
+```js
+their_events: {"ape", "bee", "cat", "doe", "eel", "fox", "gnu", "hog"}
+your_events:  {"ape", "eel", "fox", "gnu"}
+
+them: <-["ape", 0, "bee", 0, "cat", 0, "doe", 0, "eel", h("fox"), "gnu", 0, "hog", 0 "ape"]-
+```
+
+You now have the same hash between `"ape"`, `"bee"`, `"cat"`, `"doe"`, `"eel"`, `"hog"`, and `"gnu"` there is nothing left to send.
+You send the hash covering the entire range so they know you are done and can restart if the sets have diverged as the protocal was running.
+```js
+their_events: {"ape", "bee", "cat", "doe", "eel", "fox", "gnu", "hog"}
+your_events:  {"ape", "bee", "cat", "doe", "eel", "fox", "gnu", "hog"}
+
+you: -["ape", h("bee", "cat", "doe", "eel", "fox", "hog", "gnu"), "ape"]->
+
+```
+
+In this example only `"fox"` was never sent across the wire but in large sets with only tiny fractions of diffrences most keys will never be sent across the wire.
+> ToDo: should we merge 
+> ```
+> ["ape", h("bee", "cat", "doe", "eel", "fox", "hog", "gnu"), "ape"]
+> ```
+>  to
+> ```
+> [h("ape", "bee", "cat", "doe", "eel", "fox", "hog", "gnu")]
+> ```
+> to represent the entire ring?
+
+---
+#### **Imutable object store**
```

Comment:
This section is a bit too abstract for me to fully follow what you are getting at.